File: public/java/024-remove-nth-node-from-end.java
/*
 * LeetCode Problem #19: Remove Nth Node From End of List
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/remove-nth-node-from-end-of-list/
 */
import java.util.*;

class Solution {

    // Definition for singly-linked list.
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int val) { this.val = val; }
    }

    // APPROACH 1: Two Pointer One Pass | O(n) time | O(1) space
    // EXPLAIN: Advance fast pointer n+1 steps ahead; then move both until fast reaches the end — slow is at the target's predecessor.
    // WHEN: Optimal — single pass, constant space; classic two-pointer linked list trick.
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy;
        ListNode slow = dummy;

        // Advance fast by n+1 steps
        for (int i = 0; i <= n; i++) {
            fast = fast.next;
        }

        // Move both until fast reaches end
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }

        slow.next = slow.next.next;
        return dummy.next;
    }

    // APPROACH 2: Two Pass Length | O(n) time | O(1) space
    // EXPLAIN: First pass counts the length; second pass navigates to the (L - n)th node and removes the next.
    // WHEN: Simple to explain; good when two passes are acceptable.
    public ListNode removeNthFromEnd_TwoPass(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        // Count length
        int length = 0;
        ListNode temp = head;
        while (temp != null) {
            length++;
            temp = temp.next;
        }

        // Navigate to node just before the target
        int stepsFromHead = length - n;
        ListNode current = dummy;
        for (int i = 0; i < stepsFromHead; i++) {
            current = current.next;
        }
        current.next = current.next.next;

        return dummy.next;
    }

    // APPROACH 3: Recursion | O(n) time | O(n) space
    // EXPLAIN: Recurse to the end; count nodes on the return path and skip the nth node from the end.
    // WHEN: When a recursive solution is preferred; stack depth is O(n).
    public ListNode removeNthFromEnd_Recursive(ListNode head, int n) {
        int[] count = {0};
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        dummy.next = removeHelper(head, n, count);
        return dummy.next;
    }

    private ListNode removeHelper(ListNode node, int n, int[] count) {
        if (node == null) return null;
        node.next = removeHelper(node.next, n, count);
        count[0]++;
        if (count[0] == n) return node.next; // skip this node
        return node;
    }

    // APPROACH 4: Stack | O(n) time | O(n) space
    // EXPLAIN: Push all nodes (including dummy) onto a stack; pop n times to reach the predecessor, then unlink the target.
    // WHEN: Intuitive LIFO-based approach for accessing nodes from the end.
    public ListNode removeNthFromEnd_Stack(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        Deque<ListNode> stack = new ArrayDeque<>();
        ListNode current = dummy;
        while (current != null) {
            stack.push(current);
            current = current.next;
        }

        // Pop n nodes to reach the predecessor
        for (int i = 0; i < n; i++) {
            stack.pop();
        }

        ListNode prev = stack.peek();
        prev.next = prev.next.next;

        return dummy.next;
    }
}

// Made with Bob

File: public/python/024-remove-nth-node-from-end.py
# LeetCode Problem #19: Remove Nth Node From End of List
# Difficulty: Medium
# Link: https://leetcode.com/problems/remove-nth-node-from-end-of-list/

from __future__ import annotations
from typing import Optional

# ─────────────────────────────────────────────
# Shared node definition
class ListNode:
    def __init__(self, val: int = 0, nxt: Optional["ListNode"] = None) -> None:
        self.val  = val
        self.next = nxt


# ─────────────────────────────────────────────
# APPROACH 1: Two Pointer One Pass | O(n) time | O(1) space
# EXPLAIN: Advance the fast pointer n+1 steps ahead; move both until fast reaches None — slow is at the predecessor.
# WHEN: Elegant single-pass solution; the go-to answer in interviews.

def remove_nth_one_pass(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    dummy = ListNode(0, head)
    slow = fast = dummy
    # Advance fast by n+1 steps so the gap is n
    for _ in range(n + 1):
        fast = fast.next        # type: ignore[assignment]
    while fast is not None:
        slow = slow.next        # type: ignore[assignment]
        fast = fast.next
    slow.next = slow.next.next  # type: ignore[union-attr]
    return dummy.next


# ─────────────────────────────────────────────
# APPROACH 2: Two Pass Length | O(n) time | O(1) space
# EXPLAIN: First pass counts length L; second pass stops at node L - n to remove the target.
# WHEN: Simple and explicit — easy to verify correctness; identical time complexity to one pass.

def remove_nth_two_pass(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    dummy = ListNode(0, head)
    # Count length
    length, cur = 0, head
    while cur:
        length += 1
        cur     = cur.next
    # Move to the node just before the target
    target = length - n
    cur = dummy
    for _ in range(target):
        cur = cur.next          # type: ignore[assignment]
    cur.next = cur.next.next    # type: ignore[union-attr]
    return dummy.next


# ─────────────────────────────────────────────
# APPROACH 3: Recursion | O(n) time | O(n) space
# EXPLAIN: Recurse to the end; count nodes on the return path and skip the nth one from the end.
# WHEN: When a recursive solution is preferred; call-stack depth is O(n).

def remove_nth_recursive(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    count = [0]   # mutable counter via list

    def helper(node: Optional[ListNode]) -> Optional[ListNode]:
        if node is None:
            return None
        node.next = helper(node.next)
        count[0] += 1
        if count[0] == n:
            return node.next   # skip this node
        return node

    dummy = ListNode(0, head)
    dummy.next = helper(head)
    return dummy.next


# ─────────────────────────────────────────────
# APPROACH 4: Stack | O(n) time | O(n) space
# EXPLAIN: Push all nodes onto a stack; pop n times to reach the predecessor of the target, then unlink it.
# WHEN: When a stack-based approach is natural or required.

def remove_nth_stack(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    dummy = ListNode(0, head)
    stack: list[ListNode] = []
    cur: Optional[ListNode] = dummy
    while cur:
        stack.append(cur)
        cur = cur.next
    # Pop n nodes to reach the predecessor
    for _ in range(n):
        stack.pop()
    prev = stack[-1]
    prev.next = prev.next.next  # type: ignore[union-attr]
    return dummy.next


# ─────────────────────────────────────────────
# Helpers
def build_list(vals: list[int]) -> Optional[ListNode]:
    dummy = ListNode()
    cur   = dummy
    for v in vals:
        cur.next = ListNode(v)
        cur       = cur.next
    return dummy.next

def to_list(head: Optional[ListNode]) -> list[int]:
    result, cur = [], head
    while cur:
        result.append(cur.val)
        cur = cur.next
    return result

if __name__ == "__main__":
    cases = [
        ([1,2,3,4,5], 2, [1,2,3,5]),
        ([1],          1, []),
        ([1,2],        1, [1]),
        ([1,2],        2, [2]),
    ]
    for vals, n, expected in cases:
        assert to_list(remove_nth_one_pass( build_list(vals), n)) == expected
        assert to_list(remove_nth_two_pass( build_list(vals), n)) == expected
        assert to_list(remove_nth_recursive(build_list(vals), n)) == expected
        assert to_list(remove_nth_stack(    build_list(vals), n)) == expected
    print("All tests passed.")

# Made with Bob
